File: results/2026-07-01_184135/tasks/16-environment-matrix-generator/typescript-bun-fable5-medium/generated-code/scripts/run-act-tests.ts
/**
 * End-to-end act test harness.
 *
 * Every test case is executed THROUGH the GitHub Actions workflow via
 * `act push --rm`, never by calling the script directly:
 *
 *   1. For each case: build a temp git repo containing the project files
 *      plus that case's fixture as config.json (and an EXPECT_FAILURE
 *      marker for the error-path case), commit, and run `act push --rm`.
 *   2. Append each case's full act output, clearly delimited, to
 *      act-result.txt in the project root.
 *   3. Assert act exited 0 for every case.
 *   4. Assert the output contains the EXACT expected matrix JSON (or the
 *      exact validation error) for that case's input — known-good values
 *      hardcoded below, not merely "some output appeared".
 *   5. Assert both jobs ("Unit tests" and "Generate build matrix")
 *      report "Job succeeded".
 *
 * Run with: bun run scripts/run-act-tests.ts
 */
import { appendFileSync, cpSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const ROOT = join(import.meta.dir, "..");
const RESULT_FILE = join(ROOT, "act-result.txt");

/** Files/dirs that make up the project inside each temp repo. */
const PROJECT_FILES = [
  ".github",
  ".actrc",
  "src",
  "tests",
  "fixtures",
  "package.json",
  "tsconfig.json",
];

interface TestCase {
  name: string;
  fixture: string;
  /** Create the EXPECT_FAILURE marker consumed by the workflow. */
  expectFailure?: boolean;
  /** Exact substrings that MUST appear in the act output. */
  expectedInOutput: string[];
}

// Known-good expected values, derived from each fixture's input by hand:
// case 1: 2 OS x 2 versions x 1 feature = exactly these 4 combinations.
// case 2: 4 base combos - 1 excluded (macos/18) + 1 included (ubuntu/22/flagB) = 4.
// case 3: 3 x 4 x 3 = 36 combos > max-size 10 -> exact validation error.
const CASES: TestCase[] = [
  {
    name: "case1-basic",
    fixture: "fixtures/case1-basic.json",
    expectedInOutput: [
      'MATRIX_JSON={"strategy":{"fail-fast":true,"max-parallel":2,"matrix":{"os":["ubuntu-latest","macos-latest"],"version":["18","20"],"feature":["stable"]}},"combinations":[{"os":"ubuntu-latest","version":"18","feature":"stable"},{"os":"ubuntu-latest","version":"20","feature":"stable"},{"os":"macos-latest","version":"18","feature":"stable"},{"os":"macos-latest","version":"20","feature":"stable"}],"count":4}',
    ],
  },
  {
    name: "case2-include-exclude",
    fixture: "fixtures/case2-include-exclude.json",
    expectedInOutput: [
      'MATRIX_JSON={"strategy":{"fail-fast":false,"matrix":{"os":["ubuntu-latest","macos-latest"],"version":["18","20"],"feature":["flagA"],"include":[{"os":"ubuntu-latest","version":"22","feature":"flagB"}],"exclude":[{"os":"macos-latest","version":"18"}]}},"combinations":[{"os":"ubuntu-latest","version":"18","feature":"flagA"},{"os":"ubuntu-latest","version":"20","feature":"flagA"},{"os":"macos-latest","version":"20","feature":"flagA"},{"os":"ubuntu-latest","version":"22","feature":"flagB"}],"count":4}',
    ],
  },
  {
    name: "case3-oversized-error",
    fixture: "fixtures/case3-oversized.json",
    expectFailure: true,
    expectedInOutput: [
      "MATRIX_ERROR_CONFIRMED: error: Matrix size 36 exceeds maximum allowed size 10",
    ],
  },
];

/** Both workflow jobs must report success in the act log. */
const JOB_SUCCESS_MARKERS = [
  "[Environment Matrix Generator/Unit tests]",
  "[Environment Matrix Generator/Generate build matrix]",
];

function sh(cmd: string[], cwd: string): { code: number; output: string } {
  const proc = Bun.spawnSync(cmd, { cwd, stdout: "pipe", stderr: "pipe" });
  return {
    code: proc.exitCode,
    output: proc.stdout.toString() + proc.stderr.toString(),
  };
}

function fail(msg: string): never {
  console.error(`\nHARNESS FAILURE: ${msg}`);
  process.exit(1);
}

// Start a fresh result file for this run.
writeFileSync(RESULT_FILE, `act test run started\n`);

let passed = 0;
for (const tc of CASES) {
  console.log(`\n=== ${tc.name} ===`);
  const repo = mkdtempSync(join(tmpdir(), `matrix-act-${tc.name}-`));
  try {
    // 1. Assemble the temp repo: project files + this case's config.
    for (const f of PROJECT_FILES) {
      cpSync(join(ROOT, f), join(repo, f), { recursive: true });
    }
    cpSync(join(ROOT, tc.fixture), join(repo, "config.json"));
    if (tc.expectFailure) {
      writeFileSync(join(repo, "EXPECT_FAILURE"), "expect CLI failure\n");
    }
    for (const git of [
      ["git", "init", "-q", "-b", "main"],
      ["git", "config", "user.email", "ci@example.com"],
      ["git", "config", "user.name", "CI"],
      ["git", "add", "-A"],
      ["git", "commit", "-q", "-m", `fixture ${tc.name}`],
    ]) {
      const r = sh(git, repo);
      if (r.code !== 0) fail(`git setup failed for ${tc.name}: ${r.output}`);
    }

    // 2. Run the workflow through act and record the full output.
    console.log("running act push --rm ...");
    // --pull=false: the runner image is local-only (mapped in .actrc);
    // pulling would fail against the public registry.
    const act = sh(["act", "push", "--rm", "--pull=false"], repo);
    appendFileSync(
      RESULT_FILE,
      `\n${"=".repeat(72)}\n=== TEST CASE: ${tc.name} (fixture: ${tc.fixture}) ===\n` +
        `=== act exit code: ${act.code} ===\n${"=".repeat(72)}\n${act.output}\n`,
    );

    // 3. act itself must succeed.
    if (act.code !== 0) {
      fail(`act exited with code ${act.code} for ${tc.name} (see act-result.txt)`);
    }
    // 4. Exact expected values for this case's input.
    for (const expected of tc.expectedInOutput) {
      if (!act.output.includes(expected)) {
        fail(`${tc.name}: expected exact output not found:\n  ${expected}`);
      }
    }
    // 5. Every job must report success.
    for (const jobPrefix of JOB_SUCCESS_MARKERS) {
      const jobSucceeded = act.output
        .split("\n")
        .some((l) => l.includes(jobPrefix) && l.includes("Job succeeded"));
      if (!jobSucceeded) {
        fail(`${tc.name}: no "Job succeeded" line for ${jobPrefix}`);
      }
    }
    console.log(`PASS: ${tc.name}`);
    passed++;
  } finally {
    rmSync(repo, { recursive: true, force: true });
  }
}

appendFileSync(RESULT_FILE, `\nALL ${passed}/${CASES.length} ACT TEST CASES PASSED\n`);
console.log(`\nALL ${passed}/${CASES.length} ACT TEST CASES PASSED (details: act-result.txt)`);

File: results/2026-06-30_191904/tasks/16-environment-matrix-generator/default-sonnet5-1m-medium/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""
Act-based integration test harness.

Per project requirements, every test case must be exercised end-to-end
through the real GitHub Actions workflow via `act`, not by calling
matrix_generator.py directly. For each test case this script:

  1. Creates an isolated temp directory containing a copy of the project
     (script, tests, workflow, fixtures) plus that case's fixture data for
     fixtures/ci_matrix.json (the file the workflow's generate-matrix job
     reads).
  2. Initializes a throwaway git repo there (act needs a git repo) and
     commits the files.
  3. Runs `act push --rm` inside that repo and captures combined output.
  4. Appends the output to act-result.txt (in the original working
     directory), clearly delimited per test case.
  5. Asserts exit code 0, asserts every job reports "Job succeeded", and
     asserts the output contains the exact expected matrix values for that
     case's input config.

Run with: python3 run_act_tests.py
"""
import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).parent.resolve()
RESULT_FILE = ROOT / "act-result.txt"

PROJECT_FILES = [
    "matrix_generator.py",
    "test_matrix_generator.py",
    "test_workflow_structure.py",
    ".actrc",
]
PROJECT_DIRS = [
    "fixtures",
    ".github",
]

# Each case overrides fixtures/ci_matrix.json (the file the workflow's
# generate-matrix job actually reads) with different config, then asserts
# the exact resulting matrix/fail-fast/max-parallel values appear in the
# act output.
CASES = [
    {
        "name": "two-python-versions-fail-fast",
        "config": {
            "dimensions": {
                "os": ["ubuntu-latest"],
                "python-version": ["3.10", "3.11"],
            },
            "fail_fast": True,
            "max_parallel": 2,
            "max_size": 10,
        },
        "expected_strategy_json": json.dumps(
            {
                "fail-fast": True,
                "max-parallel": 2,
                "matrix": {
                    "include": [
                        {"os": "ubuntu-latest", "python-version": "3.10"},
                        {"os": "ubuntu-latest", "python-version": "3.11"},
                    ]
                },
            }
        ),
        "expected_job_count": 2,  # build (3.10), build (3.11)
    },
    {
        "name": "include-merge-three-versions",
        "config": {
            "dimensions": {
                "os": ["ubuntu-latest"],
                "python-version": ["3.10", "3.11", "3.12"],
            },
            "include": [
                {"os": "ubuntu-latest", "python-version": "3.12", "experimental": True}
            ],
            "fail_fast": False,
            "max_parallel": 2,
            "max_size": 10,
        },
        "expected_strategy_json": json.dumps(
            {
                "fail-fast": False,
                "max-parallel": 2,
                "matrix": {
                    "include": [
                        {"os": "ubuntu-latest", "python-version": "3.10"},
                        {"os": "ubuntu-latest", "python-version": "3.11"},
                        {
                            "os": "ubuntu-latest",
                            "python-version": "3.12",
                            "experimental": True,
                        },
                    ]
                },
            }
        ),
        "expected_job_count": 3,  # build (3.10), build (3.11), build (3.12)
    },
]


def setup_case_repo(tmp_dir: Path, case: dict) -> None:
    for name in PROJECT_FILES:
        shutil.copy2(ROOT / name, tmp_dir / name)
    for name in PROJECT_DIRS:
        shutil.copytree(ROOT / name, tmp_dir / name)

    # Overwrite the fixture the workflow actually reads with this case's
    # config, so a single workflow file drives every test case.
    fixture_path = tmp_dir / "fixtures" / "ci_matrix.json"
    fixture_path.write_text(json.dumps(case["config"], indent=2) + "\n")

    subprocess.run(["git", "init", "-q"], cwd=tmp_dir, check=True)
    subprocess.run(["git", "config", "user.email", "test@example.com"], cwd=tmp_dir, check=True)
    subprocess.run(["git", "config", "user.name", "Test Harness"], cwd=tmp_dir, check=True)
    subprocess.run(["git", "add", "-A"], cwd=tmp_dir, check=True)
    subprocess.run(["git", "commit", "-q", "-m", "test fixture commit"], cwd=tmp_dir, check=True)


def run_act(tmp_dir: Path) -> subprocess.CompletedProcess:
    return subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=tmp_dir,
        capture_output=True,
        text=True,
        timeout=600,
    )


def run_case(case: dict) -> None:
    print(f"=== Running act test case: {case['name']} ===")
    with tempfile.TemporaryDirectory(prefix="matrix-gen-act-") as tmp:
        tmp_dir = Path(tmp)
        setup_case_repo(tmp_dir, case)
        proc = run_act(tmp_dir)

    combined_output = proc.stdout + "\n" + proc.stderr

    with open(RESULT_FILE, "a", encoding="utf-8") as f:
        f.write(f"\n{'=' * 80}\nTEST CASE: {case['name']}\n{'=' * 80}\n")
        f.write(f"Exit code: {proc.returncode}\n\n")
        f.write(combined_output)
        f.write("\n")

    assert proc.returncode == 0, (
        f"[{case['name']}] act exited with {proc.returncode}, expected 0:\n{combined_output}"
    )

    job_success_count = combined_output.count("Job succeeded")
    # 1 unit-tests + 1 generate-matrix + N build (one per matrix entry)
    expected_success_count = 2 + case["expected_job_count"]
    assert job_success_count == expected_success_count, (
        f"[{case['name']}] expected {expected_success_count} 'Job succeeded' "
        f"occurrences, found {job_success_count}:\n{combined_output}"
    )

    assert case["expected_strategy_json"] in combined_output, (
        f"[{case['name']}] expected exact strategy JSON not found in act output.\n"
        f"Expected: {case['expected_strategy_json']}\n"
        f"Output:\n{combined_output}"
    )

    print(f"=== PASSED: {case['name']} ===\n")


def main() -> int:
    if RESULT_FILE.exists():
        RESULT_FILE.unlink()

    failures = []
    for case in CASES:
        try:
            run_case(case)
        except AssertionError as exc:
            failures.append(str(exc))
            print(f"=== FAILED: {case['name']} ===\n{exc}\n", file=sys.stderr)

    if failures:
        print(f"\n{len(failures)} of {len(CASES)} act test case(s) failed.", file=sys.stderr)
        return 1

    print(f"\nAll {len(CASES)} act test cases passed. See {RESULT_FILE}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
